okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt

```diff
@@ -0,0 +1,283 @@
+                  sleep(100)
+                }
+                fail("Expected connection to be closed")
+              }
+            })
+            .build()
+    )
+    cancelLater(call, 500)
+    try {
+      call.execute()
+      fail()
+    } catch (expected: IOException) {
+      assertEquals(cancelMode == INTERRUPT, Thread.interrupted())
+    }
+    assertEquals(if (connectionType == H2) 1 else 0, client.connectionPool.connectionCount())
+  }
+
+  @Test
+  fun cancelReadingResponseBody() {
+    val responseBodySize = 8 * 1024 * 1024 // 8 MiB.
+    server.enqueue(
+        MockResponse()
+            .setBody(
+                Buffer()
+                    .write(ByteArray(responseBodySize))
+            )
+            .throttleBody(64 * 1024.toLong(), 125, MILLISECONDS)
+    ) // 500 Kbps
+    val call = client.newCall(
+        Builder()
+            .url(server.url("/"))
+            .build()
+    )
+    val response = call.execute()
+    cancelLater(call, 500)
+    val responseBody = response.body!!.byteStream()
+    val buffer = ByteArray(1024)
+    try {
+      while (responseBody.read(buffer) != -1) {
+      }
+      fail("Expected connection to be closed")
+    } catch (expected: IOException) {
+      assertEquals(cancelMode == INTERRUPT, Thread.interrupted())
+    }
+    responseBody.close()
+    assertEquals(if (connectionType == H2) 1 else 0, client.connectionPool.connectionCount())
+  }
+
+  @Test
+  fun cancelAndFollowup() {
+    val responseBodySize = 8 * 1024 * 1024 // 8 MiB.
+    server.enqueue(
+        MockResponse()
+            .setBody(
+                Buffer()
+                    .write(ByteArray(responseBodySize))
+            )
+            .throttleBody(64 * 1024.toLong(), 125, MILLISECONDS)
+    ) // 500 Kbps
+    server.enqueue(MockResponse().apply {
+      setResponseCode(200)
+      setBody(".")
+    })
+
+    val call = client.newCall(Builder().url(server.url("/")).build())
+    val response = call.execute()
+    cancelLater(call, 500)
+    val responseBody = response.body!!.byteStream()
+    val buffer = ByteArray(1024)
+    try {
+      while (responseBody.read(buffer) != -1) {
+      }
+      fail("Expected connection to be closed")
+    } catch (expected: IOException) {
+      assertEquals(cancelMode == INTERRUPT, Thread.interrupted())
+    }
+    responseBody.close()
+    assertEquals(if (connectionType == H2) 1 else 0, client.connectionPool.connectionCount())
+
+    val call2 = client.newCall(Builder().url(server.url("/")).build())
+    call2.execute().use {
+      assertEquals(".", it.body!!.string())
+    }
+
+    val events = listener.eventSequence.filter { isConnectionEvent(it) }.map { it.name }
+
+    val expectedEvents = when {
+      cancelMode == CANCEL && connectionType == H2 -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "Canceled", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
```

Comment:
build this in steps to make it more legible? As-is it’s not obvious where the deltas are.

Reply:
Good call, turned out much nicer.